plutus-core/plutus-ir/src/PlutusIR/Analysis/RetainedSize.hs

```diff
@@ -0,0 +1,215 @@
+annotateWithSizes :: DirectionRetentionMap -> Tree Int -> Tree (Int, Size)
+annotateWithSizes sizeInfo = go where
+    go (Node i ts) = Node (i, sizeI) rs where
+        rs = map go ts
+        sizeI = lookupSize i sizeInfo <> foldMap (snd . rootLabel) rs
+
+-- | Compute the dominator tree of a graph.
+toDomTree :: C.Graph Node -> Tree Int
+toDomTree = domTree . (,) rootInt . adjacencyIntMap . fmap nodeToInt
+
+-- | Compute the retention map of a graph.
+depsRetentionMap :: DirectionRetentionMap -> C.Graph Node -> IntMap Size
+depsRetentionMap sizeInfo = IntMap.fromList . flatten . annotateWithSizes sizeInfo . toDomTree
+
+-- | Construct a 'UniqueMap' having size information for each individual part of a 'Binding'.
+bindingSize
+    :: (HasUnique tyname TypeUnique, HasUnique name TermUnique)
+    => Binding tyname name uni fun ann -> UniqueMap Unique Size
+bindingSize (TermBind _ _ var term) =
+    insertByNameIndex var (varDeclSize var <> termSize term) mempty
+bindingSize (TypeBind _ tyVar ty) =
+    insertByNameIndex tyVar (tyVarDeclSize tyVar <> typeSize ty) mempty
+bindingSize (DatatypeBind _ (Datatype _ dataDecl params matchName constrs))
+    = insertByNameIndex dataDecl (tyVarDeclSize dataDecl)
+    . flip (foldr $ \param -> insertByNameIndex param $ tyVarDeclSize param) params
+    . insertByNameIndex matchName (Size 1)
+    . flip (foldr $ \constr -> insertByNameIndex constr $ varDeclSize constr) constrs
+    $ mempty
+
+-- | Construct a 'UniqueMap' having size information for each individual part of every 'Binding'
+-- in a term.
+bindingSizes
+    :: (HasUnique tyname TypeUnique, HasUnique name TermUnique)
+    => Term tyname name uni fun ann -> UniqueMap Unique Size
+bindingSizes (Let _ _ binds term) = foldMap bindingSize binds <> bindingSizes term
+bindingSizes term                 = term ^. termSubterms . to bindingSizes
+
+-- | Same as 'bindingSizes' but is wrapped in a newtype and has a bogus entry for the root.
+toDirectionRetentionMap
+    :: (HasUnique tyname TypeUnique, HasUnique name TermUnique)
+    => Term tyname name uni fun ann -> DirectionRetentionMap
+toDirectionRetentionMap term =
+    DirectionRetentionMap . IntMap.insert rootInt rootSize . unUniqueMap $ bindingSizes term where
+        -- See Note [Handling the root].
+        rootSize = Size (- 10 ^ (10::Int))
+
+-- | Check if a 'Node' appears in 'DirectionRetentionMap'.
+hasSizeIn :: DirectionRetentionMap -> Node -> Bool
+hasSizeIn _             Root                                   = True
+hasSizeIn (DirectionRetentionMap ss) (Variable (PLC.Unique i)) = i `IntMap.member` ss
+
+-- | Compute the retention map of a term.
+termRetentionMap
+    :: (HasUnique tyname TypeUnique, HasUnique name TermUnique, ToBuiltinMeaning uni fun)
+    => Term tyname name uni fun ann -> IntMap Size
+termRetentionMap term = depsRetentionMap sizeInfo deps where
+    sizeInfo = toDirectionRetentionMap term
+    deps = C.induce (hasSizeIn sizeInfo) $ fst $ runTermDeps term
```

Comment:
Given that you do this before computing dominators, I think you're relying on the fact that nodes that don't retain any size are also not interesting for dominator analysis. But that seems like it could give us wrong results if it breaks a chain of dependencies. Maybe I need to think of an example.

Reply:
Great catch. That sounds like it might be troubling indeed, but I think it's also possible that this is completely fine. An example is needed. For the time being I added a comment.